IdentityCore/src/network/MSIDHttpRequest.m

```diff
@@ -0,0 +1,128 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDHttpRequest.h"
+#import "MSIDJsonResponseSerializer.h"
+#import "MSIDUrlRequestSerializer.h"
+#import "MSIDHttpRequestTelemetryProtocol.h"
+
+@interface MSIDHttpRequest () <NSURLSessionDelegate>
+
+@property (nonatomic) NSURLSessionConfiguration *sessionConfiguration;
+@property (nonatomic) NSOperationQueue *operationQueue;
+@property (nonatomic) NSURLSession *session;
+
+@end
+
+@implementation MSIDHttpRequest
+
+- (instancetype)init
+{
+    self = [super init];
+    
+    if (!self) return nil;
+    
+    // TODO: can we remove queue?
+    _operationQueue = [NSOperationQueue new];
+    _operationQueue.maxConcurrentOperationCount = 1;
+    
+    _sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
+    _session = [NSURLSession sessionWithConfiguration:_sessionConfiguration delegate:self delegateQueue:self.operationQueue];
+    _responseSerializer = [MSIDJsonResponseSerializer new];
+    _requestSerializer = [MSIDUrlRequestSerializer new];
+    
+    _retryOnErrorCounter = 1;
+    
+    return self;
+}
+
+- (NSURLRequest *)urlRequest
+{
+    if (!_urlRequest)
+    {
+        __auto_type request = [NSMutableURLRequest new];
+        request.timeoutInterval = 300;
+        [request setValue:@"application/json" forHTTPHeaderField:@"Accept"];
+        
+        _urlRequest = request;
+    }
+    
+    return _urlRequest;
+}
+
+- (void)sendWithBlock:(MSIDHttpRequestDidCompleteBlock _Nullable )completionBlock;
+{
+    self.urlRequest = [self.requestSerializer serializeWithRequest:self.urlRequest parameters:self.parameters];
+    
+    [self.telemetry sendRequestEventWithId:self.context.telemetryRequestId];
+    
+    [[self.session dataTaskWithRequest:self.urlRequest completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error)
+      {
+          MSID_LOG_VERBOSE(self.context, @"Received network response: %@, error %@", _PII_NULLIFY(response), _PII_NULLIFY(error));
+          MSID_LOG_VERBOSE_PII(self.context, @"Received network response: %@, error %@", response, error);
+          
+          [self.telemetry responseReceivedEventWithId:self.context.telemetryRequestId
+                                        correlationId:self.context.correlationId
+                                           urlRequest:self.urlRequest
+                                          urlResponse:response
+                                                 data:data
+                                                error:error];
+          
+          if (error)
+          {
+              __auto_type httpResponse = (NSHTTPURLResponse *)response;
+              BOOL shouldRetry = [httpResponse isKindOfClass:NSHTTPURLResponse.class];
+              shouldRetry &= self.retryOnErrorCounter > 0;
+              shouldRetry &= httpResponse.statusCode >= 500 && httpResponse.statusCode <= 599;
```

Comment:
In general I agree that sending POST requests might be unsafe because POST is not idempotent. Not sure if we should worry about it in our case. We can put error counter to 0 by default, and change it to 1 only for requests that can be re-sent.